### ISSUE TITLE
Incorrect return type in overloaded functions with Literal arguments in LSP mode

### ISSUE THREAD

**Issue by vemel** [microsoft/pyright#1256]
**Describe the bug**

If `pyright` is used as an LSP, e.g. in https://github.com/emacs-lsp/lsp-pyright/, function overloads with `Literal` arguments return incorrect type. It affect methods suggestions as well, but I believe that the main reason is a wrong return type. This case works as expected in VSCode (Pylance).

**To Reproduce**

- Create a function with two `Literal` overloads that return different types.
- Call the function with Literal that returns any type other than the first one (see example below) 
- Returned result has type and method suggestions as if it was the first overload variant return type

**Expected behavior**

Overloaded functions return types to have correct method suggestions.

**Code**

```python
from datetime import datetime
from typing import Literal, Union, overload


@overload
def my_func(value: Literal["date"]) -> datetime:
    ...


@overload
def my_func(value: Literal["int"]) -> int:
    ...


def my_func(value: str) -> Union[datetime, int]:
    return datetime.now() if value == "date" else 1


my_date = my_func("date")  # my_date has a correct type (datetime) and methods auto-complete works as expected
my_int = my_func("int")  # my_int uses incorrect type (datetime) and methods auto-complete works as it is a datetime
```

**VS Code extension or command-line**

VSCode (Pylance) handles this case correctly, but https://github.com/emacs-lsp/lsp-pyright/blob/master/lsp-pyright.el and https://github.com/sublimelsp/LSP-pyright are affected. Maybe it is a wrong configuration in run arguments.

**Additional context**

Please see the original issue report here: https://github.com/vemel/mypy_boto3_builder/issues/59

**Thank you**

This is the best type checker I have ever used, keep up a good work!

**Comment by erictraut** [comment]
I'm not able to repro the problem with the code you've provided. Here's what I see when I add some calls to `reveal_type`. It looks correct to me. Are you seeing different results?

![Screen Shot 2020-12-09 at 9 24 16 AM](https://user-images.githubusercontent.com/7040122/101664114-50816380-3a00-11eb-8185-696590a0cbe9.png)

**Comment by erictraut** [comment]
Incidentally, you shouldn't see any behavioral differences between the VS Code extension, the CLI, or the LSP. They all use the same code for type analysis.

**Comment by rirze** [comment]
@vemel took the liberty of posting this question/bug report here on my behalf. In my original filing, I noted that certain attributes were not showing as autocomplete options. Using the example code above, when I type 
```python3
my_int = my_func('int')
my_int.
```
and hover at the autocomplete at `my_int.`, I'm seeing `datetime` and `int` attributes. Meaning something is not properly following the overloads to restrict the typing to just `int` attributes. 

Here's what the autocomplete gives me at that point:
<details>
  <summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/3255334/101671559-a2d37c00-3a22-11eb-8088-1e8ff7c493dc.png)
</details>

Again this is for lsp-pyright using Emacs. I'm not sure if this is intended behavior or differing behavior. Clarifying that would help.

**Comment by vemel** [comment]
@erictraut , the example works great in VSCode, indeed. I am not fully sure how it is different from using pyright as LSP in SublimeText or Emacs, but obviously there is a difference. 

I could not reproduce this issue in VSCode as well. Sorry if my bug report is not clear.

**Comment by erictraut** [comment]
If your editor supports hover text, could you please hover over the `my_int` symbol to see what type Pyright thinks it is? Alternatively, insert a call to `reveal_type(my_int)` between the two lines above, and see what type it shows.

As I mentioned, the core type logic is the same between the Pyright VS Code extension, the CLI and the LSP. If you are seeing difference in behavior, it is either because of the client or a configuration difference. I can't think of any configuration settings that would affect the behavior in this manner, so my best guess is that it's a problem with the client.

**Comment by rirze** [comment]
Seems like `reveal_type` is working as expected?
![image](https://user-images.githubusercontent.com/3255334/101672633-1f1a8f00-3a24-11eb-8a91-ef0bfc844490.png)

If you think it could be client/configuration related, would you have any suggestions for what I could try?

**Comment by erictraut** [comment]
OK, so Pyright is evaluating the correct type.

You would need to work with the maintainers of the client (or the LSP mechanism for your client) to debug the issue. It sounds like it is displaying something other than the completion suggestions returned by the Pyright LSP.

Another solution is to switch to VS Code as your editor. :)

**Comment by rirze** [comment]
This is definitely a problem with my setup and not any projects. Creating a simplified config file results in expected behavior. 

I have nothing to add, so this issue can closed from my end.
